### ISSUE TITLE
TimeOut attribute doesn't work with TestCaseSource Attribute

### ISSUE THREAD

**Issue by GibbOne** [nunit/nunit#755]
I'm using NUnit version 3.0.0-beta-3 and the dot net Framework 4.6.

Example:

```C#
using NUnit.Framework;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace ClassLibrary1
{
    public class TestTimeOut
    {
        // all in ms
        const int TIME_OUT_TIME = 2000;
        const int NOT_TIMEOUTED_TIME = 1000;
        const int TIMEOUTED_TIME = 3000;

        // it works!
        [Test, Timeout(TIME_OUT_TIME), Parallelizable(ParallelScope.Fixtures)]
        public async Task TestTimeOutElapsed()
        {
            await TestTimeOutCase(TIMEOUTED_TIME);
        }

        // it works!
        [Test, Timeout(TIME_OUT_TIME), Parallelizable(ParallelScope.Fixtures)]
        public async Task TestTimeOutNotElapsed()
        {
            await TestTimeOutCase(NOT_TIMEOUTED_TIME);
        }

        // it doesn't work (both tests end successfully)
        static List<object[]> Delays = new List<object[]>() { new object[] { NOT_TIMEOUTED_TIME }, new object[] { TIMEOUTED_TIME } };

        [Test, Timeout(TIME_OUT_TIME), TestCaseSource("Delays"), Parallelizable(ParallelScope.Fixtures)]
        public async Task TestTimeOutCase(int delay)
        {
            await Task.Delay(delay);
        }

    }
}
```

Am I doing something wrong?

**Comment by constructor-igor** [comment]
![image](https://cloud.githubusercontent.com/assets/1849690/9155405/4e7d5832-3ec0-11e5-998b-3687c300f817.png)

```
<test-suite type="TestFixture" id="0-1024" name="NUnitIssue755" fullname="NUnit_v3_samples.NUnitIssue755" classname="NUnit_v3_samples.NUnitIssue755" runstate="Runnable" testcasecount="4" result="Failed" site="Child" start-time="2015-08-09 14:58:33Z" end-time="2015-08-09 14:58:40Z" duration="7.117963" total="4" passed="3" failed="1" inconclusive="0" skipped="0" asserts="0">
        <failure>
          <message>One or more child tests had errors</message>
        </failure>
        <test-suite type="ParameterizedMethod" id="0-1029" name="TestTimeOutCase" fullname="NUnit_v3_samples.NUnitIssue755.TestTimeOutCase" runstate="Runnable" testcasecount="2" result="Passed" start-time="2015-08-09 14:58:33Z" end-time="2015-08-09 14:58:37Z" duration="4.061851" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0">
          <properties>
            <property name="ParallelScope" value="Fixtures" />
            <property name="Timeout" value="2000" />
          </properties>
          <test-case id="0-1027" name="TestTimeOutCase(1000)" fullname="NUnit_v3_samples.NUnitIssue755.TestTimeOutCase(1000)" methodname="TestTimeOutCase" classname="NUnit_v3_samples.NUnitIssue755" runstate="Runnable" seed="1694934116" result="Passed" start-time="2015-08-09 14:58:33Z" end-time="2015-08-09 14:58:34Z" duration="1.031007" asserts="0" />
          <test-case id="0-1028" name="TestTimeOutCase(3000)" fullname="NUnit_v3_samples.NUnitIssue755.TestTimeOutCase(3000)" methodname="TestTimeOutCase" classname="NUnit_v3_samples.NUnitIssue755" runstate="Runnable" seed="1191636085" result="Passed" start-time="2015-08-09 14:58:34Z" end-time="2015-08-09 14:58:37Z" duration="3.005399" asserts="0" />
        </test-suite>
        <test-case id="0-1025" name="TestTimeOutElapsed" fullname="NUnit_v3_samples.NUnitIssue755.TestTimeOutElapsed" methodname="TestTimeOutElapsed" classname="NUnit_v3_samples.NUnitIssue755" runstate="Runnable" seed="1580158007" result="Failed" start-time="2015-08-09 14:58:37Z" end-time="2015-08-09 14:58:39Z" duration="2.019215" asserts="0">
          <properties>
            <property name="ParallelScope" value="Fixtures" />
            <property name="Timeout" value="2000" />
          </properties>
          <failure>
            <message>Test exceeded Timeout value of 2000ms</message>
          </failure>
        </test-case>
        <test-case id="0-1026" name="TestTimeOutNotElapsed" fullname="NUnit_v3_samples.NUnitIssue755.TestTimeOutNotElapsed" methodname="TestTimeOutNotElapsed" classname="NUnit_v3_samples.NUnitIssue755" runstate="Runnable" seed="1931312539" result="Passed" start-time="2015-08-09 14:58:39Z" end-time="2015-08-09 14:58:40Z" duration="1.016303" asserts="0">
          <properties>
            <property name="ParallelScope" value="Fixtures" />
            <property name="Timeout" value="2000" />
          </properties>
        </test-case>
      </test-suite>
```

**Comment by constructor-igor** [comment]
when I remove "async"\"await":
![image](https://cloud.githubusercontent.com/assets/1849690/9155423/075cfd9e-3ec1-11e5-9949-4fccb783ee71.png)

**Comment by CharliePoole** [comment]
What did you do for the delay when you removed async and await? 

Confirmed, in any case.

**Comment by constructor-igor** [comment]
This test is expected to fall
```c#
    [TestFixture]
    public class NUnitIssue755_SingleTest
    {
        [Timeout(2000), TestCase(3000)]
        public void TestTimeOutCase_via_TestCase(int delay)
        {
            Thread.Sleep(delay);
        }
    }
```
but it passes
```xml
    <test-suite type="TestSuite" id="0-1085" name="NUnit_v3_samples" fullname="NUnit_v3_samples" runstate="Runnable" testcasecount="46" result="Passed" start-time="2015-08-09 19:11:05Z" end-time="2015-08-09 19:11:09Z" duration="3.053555" total="1" passed="1" failed="0" inconclusive="0" skipped="0" asserts="0">
      <test-suite type="TestFixture" id="0-1038" name="NUnitIssue755_SingleTest" fullname="NUnit_v3_samples.NUnitIssue755_SingleTest" classname="NUnit_v3_samples.NUnitIssue755_SingleTest" runstate="Runnable" testcasecount="1" result="Passed" start-time="2015-08-09 19:11:05Z" end-time="2015-08-09 19:11:09Z" duration="3.050323" total="1" passed="1" failed="0" inconclusive="0" skipped="0" asserts="0">
        <test-suite type="ParameterizedMethod" id="0-1040" name="TestTimeOutCase_via_TestCase" fullname="NUnit_v3_samples.NUnitIssue755_SingleTest.TestTimeOutCase_via_TestCase" runstate="Runnable" testcasecount="1" result="Passed" start-time="2015-08-09 19:11:05Z" end-time="2015-08-09 19:11:09Z" duration="3.044782" total="1" passed="1" failed="0" inconclusive="0" skipped="0" asserts="0">
          <properties>
            <property name="Timeout" value="2000" />
          </properties>
          <test-case id="0-1039" name="TestTimeOutCase_via_TestCase(3000)" fullname="NUnit_v3_samples.NUnitIssue755_SingleTest.TestTimeOutCase_via_TestCase(3000)" methodname="TestTimeOutCase_via_TestCase" classname="NUnit_v3_samples.NUnitIssue755_SingleTest" runstate="Runnable" seed="1069402004" result="Passed" start-time="2015-08-09 19:11:06Z" end-time="2015-08-09 19:11:09Z" duration="3.013928" asserts="0" />
        </test-suite>
      </test-suite>
    </test-suite>
```

**Comment by constructor-igor** [comment]
Interesting result of next test:
```c#
    [TestFixture]
    public class NUnitIssue755_SingleTest
    {
        [Timeout(2000), Test]
        public void TestTimeOutCase_via_TestCase()
        {
            Thread.Sleep(3000);
        }
    }
```

![image](https://cloud.githubusercontent.com/assets/1849690/9156490/7ba48ea6-3ee4-11e5-8909-5517c7650bc8.png)

**Comment by CharliePoole** [comment]
@constructor-igor Do you feel like fixing this one? My guess is that parameterized tests, which are suites, are not being handled correctly with respect to this Attribute. What should be happening is that the default timeout in the TestExecutionContext gets updated by the suite initialization and then used by each method. My guess is that no initialization is happening, since this is not a fixture. You can see how it should work by tracing through execution of a test fixture with Timeout attribute on it.

**Comment by CharliePoole** [comment]
The cancelled tests would suggest that _some_ of what should be happening is happening, but on the wrong level. TimedOut tests are cancelled but we should actually recognize why they were cancelled and report them as failures, not timeouts.

**Comment by constructor-igor** [comment]
@CharliePoole I'll try to check and suggest a fix in a few days. if not, I'll inform you.

**Comment by CharliePoole** [comment]
Great. Be sure to include missing test cases in our tests.

**Comment by rprouse** [comment]
Fixes with PR #770
